Native authentication example

singingwolfboy opened the pull request:
This is based on the logic in the `examples/nylas-connect/app.py` file, but cleaned up and organized a bit. I would really appreciate if someone could run through this example manually, to verify that it works properly. (I've manually tested it myself, but it's always good to get someone else to try it, as well!)

pfista commented:
Also, when testing with Python2 I received the following error:

```
2600:1010:b000:27d6:ed5d:3ec2:84d6:dfe3 - - [25/Jul/2017 14:56:56] "GET /login/google HTTP/1.1" 302 -
2600:1010:b000:27d6:ed5d:3ec2:84d6:dfe3 - - [25/Jul/2017 14:57:19] "GET /login/google/authorized?state=EG7lCvOGH73rCuipo4HwGP5ytWPiij&code=4%2FmXBq37_p2Gt6ui-X7rW1s9K_V0WW2eCTK0f-euf4A20 HTTP/1.1" 302 -
2600:1010:b000:27d6:ed5d:3ec2:84d6:dfe3 - - [25/Jul/2017 14:57:19] "GET /google/success HTTP/1.1" 200 -
[2017-07-25 14:57:28,309] ERROR in app: Exception on /nylas/connect [GET]
Traceback (most recent call last):
  File "/Users/mik/src/nylas/nylas-python-pr/examples-new/native-authentication/venv/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/mik/src/nylas/nylas-python-pr/examples-new/native-authentication/venv/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/mik/src/nylas/nylas-python-pr/examples-new/native-authentication/venv/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/mik/src/nylas/nylas-python-pr/examples-new/native-authentication/venv/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/mik/src/nylas/nylas-python-pr/examples-new/native-authentication/venv/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "server.py", line 146, in pass_creds_to_nylas
    "google_refresh_token": google.token["refresh_token"],
KeyError: 'refresh_token'
2600:1010:b000:27d6:ed5d:3ec2:84d6:dfe3 - - [25/Jul/2017 14:57:28] "GET /nylas/connect HTTP/1.1" 500 -
```

singingwolfboy commented:
@pfista: The Python 3 bytes/str error is already resolved by this commit: https://github.com/nylas/nylas-python/commit/fed741393585d8c674d83ebb9af342dd1ba121c0. You'll need to run a version of the Nylas SDK that is Python 3 compatible, if you want to run these examples on Python 3!

I've added two more commits to this pull request: one to display the ngrok URL on startup, and one to address the refresh token error you discovered.

The refresh token error is caused by Google not handing out refresh tokens unless the user has explicitly consented to connect to the app, and by default, Google doesn't explicitly ask the user each and every time. If the user has already consented, Google remembers that consent and doesn't bother to ask again -- but then it also doesn't pass a refresh token, leading to the error you saw. The fix is to tell Google to _always_ explicitly ask, but that's not a great user experience, and most sites won't actually make the user connect more than once. I added guidance on how to enable that line of code if necessary. If you think it should always be enabled, let me know.
